### PR TITLE
feat: Observation, Diagnostic Report

### PR DESCRIPTION
**Pull request by akashkrishna619**
## Observation Template

image_2023-10-28_08-39-36

docs - [Observation Template]( https://frappehealth.com/docs/v15/user/manual/en/healthcare/observation-template )

## Observation

image_2023-10-28_08-43-53

docs - [Observation]( https://frappehealth.com/docs/v15/user/manual/en/healthcare/observation )

## Diagnostic Report

diagnostic_report

docs - [Diagnostic Report]( https://frappehealth.com/docs/v15/user/manual/en/healthcare/diagnostic-report )

### CODE REVIEW

File: healthcare/healthcare/doctype/diagnostic_report/diagnostic_report.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2023, healthcare and contributors
+// For license information, please see license.txt
+
+frappe.ui.form.on("Diagnostic Report", {
+	refresh: function(frm) {
+		show_diagnostic_report(frm);
+		if (!frm.doc.__islocal) {
+			frm.add_custom_button(__(`Get PDF1`), function () {
```
Comment (akurungadam) [inline healthcare/healthcare/doctype/diagnostic_report/diagnostic_report.js]:
```suggestion
			frm.add_custom_button(__(`Get PDF`), function () {
```

File: healthcare/healthcare/doctype/diagnostic_report/diagnostic_report.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2023, healthcare and contributors
+// For license information, please see license.txt
+
+frappe.ui.form.on("Diagnostic Report", {
+	refresh: function(frm) {
+		show_diagnostic_report(frm);
+		if (!frm.doc.__islocal) {
```
Comment (akurungadam) [inline healthcare/healthcare/doctype/diagnostic_report/diagnostic_report.js]:
```suggestion
		if (!frm.is_new()) {
```

File: healthcare/healthcare/doctype/diagnostic_report/diagnostic_report.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+from healthcare.healthcare.doctype.observation.observation import get_observation_details
+
+
+class DiagnosticReport(Document):
+	def validate(self):
+		self.set_reference_details()
+		self.set_age()
+		self.set_title()
+		# set_diagnostic_status(self)
```
Comment (akurungadam) [inline healthcare/healthcare/doctype/diagnostic_report/diagnostic_report.py:16]:
```suggestion
```

File: healthcare/healthcare/doctype/sample_collection/sample_collection.py [hunk 4]
```diff
@@ -3,13 +3,215 @@
...
+					component_observations = json.loads(obs.get("component_observations"))
+					if not any((comp["status"] == "Open") for comp in component_observations):
+						obs.status = "Collected"
+
+		if not any((obs.get("status") == "Open") for obs in self.observation_sample_collection):
+			self.status = "Collected"
+		else:
+			self.status = "Partly Collected"
+
+	# def before_submit(self):
```
Comment (akurungadam) [inline healthcare/healthcare/doctype/sample_collection/sample_collection.py:64]: isn't this relevant?

File: healthcare/healthcare/doctype/observation/observation.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+		observation_doc.result_data = result
+	# elif data_type in ["Quantity", "Numeric"]:
+	# 	observation_doc.result_float = result
+	elif data_type == "Text":
+		observation_doc.result_text = result
+	observation_doc.save()
+
+
+@frappe.whitelist()
+def add_observation(
```
Comment (akurungadam) [inline healthcare/healthcare/doctype/observation/observation.py]: too many args?

File: healthcare/healthcare/doctype/observation/observation.py [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+			if self.result_data and not is_numbers_with_exceptions(self.result_data):
+				frappe.throw(
+					_("Non numeric result {0} is not allowed for Permitted Data Type {1}").format(
+						frappe.bold(self.result_data), frappe.bold(self.permitted_data_type)
+					)
+				)
+
+
+@frappe.whitelist()
+def get_observation_details(docname):
```
Comment (akurungadam) [inline healthcare/healthcare/doctype/observation/observation.py:91]: can't we make this smaller? extract methods?

File: healthcare/healthcare/doctype/observation/observation.py [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+			"result_data",
+			"result_text",
+			"result_float",
+			"result_select",
+		]
+		for field in result_fields:
+			if self.get(field, None):
+				return True
+
+		# TODO: handle fields defaulting to now
```
Comment (akurungadam) [inline healthcare/healthcare/doctype/observation/observation.py:72]: let's remove these options in this iteration?

File: healthcare/healthcare/doctype/diagnostic_report/diagnostic_report.js [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+			frm.add_custom_button(__(`Get PDF1`), function () {
+				generate_pdf_with_print_format(frm)
+			})
+		}
+	},
+	before_save: function(frm) {
+		if (!frm.doc.__islocal && frm.is_dirty()) {
```
Comment (akurungadam) [inline healthcare/healthcare/doctype/diagnostic_report/diagnostic_report.js]:
```suggestion
		if (!frm.doc.__islocal && frm.is_dirty()) {
		if (!frm.is_new() && frm.is_dirty()) {
```

**Review by akurungadam — APPROVED** [review]

**Comment by akashkrishna619** [comment]
Unrelated test failure

**Pull request merged**